src/nanoFramework.IoT.Device.CodeConverter/GenericsTemplates/Span.cs

```diff
@@ -0,0 +1,186 @@
+                }
+            }
+
+            _array = array;
+            _start = start;
+            _length = length;
+        }
+
+        /// <summary>
+        /// Gets the element at the specified zero-based index.
+        /// </summary>
+        /// <param name="index">The zero-based index of the element.</param>
+        /// <returns>The element at the specified index.</returns>
+        // public ref T this[int index] => ref _array[_start + index]; // <= this is not working and raises exception after few access
+        public T this[int index]
+        {
+            get
+            {
+                int realIndex = _start + index;
+
+                if (realIndex >= _length)
+                {
+                    throw new ArgumentOutOfRangeException();
+                }
+
+                return _array[realIndex];
+            }
+
+            set
+            {
+                int realIndex = _start + index;
+
+                if (realIndex >= _length)
+                {
+                    throw new ArgumentOutOfRangeException();
+                }
+
+                _array[realIndex] = value;
+            }
+        }
+
+        /// <summary>
+        /// Returns an empty System.Span object.
+        /// </summary>
+        public static Span<T> Empty => new Span<T>();
+
+        /// <summary>
+        /// Returns the length of the current span.
+        /// </summary>
+        public int Length => _length;
+
+        /// <summary>
+        /// Returns a value that indicates whether the current System.Span is empty.
+        /// true if the current span is empty; otherwise, false.
+        /// </summary>
+        public bool IsEmpty => _length == 0;
+
+        /// <summary>
+        /// Copies the contents of this System.Span into a destination System.Span.
+        /// </summary>
+        /// <param name="destination"> The destination System.Span object.</param>
+        /// <exception cref="System.ArgumentException">
+        /// destination is shorter than the source System.Span.
+        /// </exception>
+        public void CopyTo(Span<T> destination)
+        {
+            if (destination.Length < _length)
+            {
+                throw new ArgumentException();
+            }
+
+            for (int i = 0; i < _length; i++)
+            {
+                destination[i] = _array[_start + i];
+            }
+        }
+
+        /// <summary>
+        /// Forms a slice out of the current span that begins at a specified index.
+        /// </summary>
+        /// <param name="start">The index at which to begin the slice.</param>
+        /// <returns>A span that consists of all elements of the current span from start to the end of the span.</returns>
+        /// <exception cref="System.ArgumentOutOfRangeException">start is less than zero or greater than System.Span.Length.</exception>
+        public Span<T> Slice(int start)
+        {
+            return Slice(start, _length - start);
```

Comment:
missing the check for the length
```
if ((start > _length) || (start < 0))
{
throw new ArgumentOutOfRangeException();
}

return new SpanByte(_array, _start + start, _length - start);
```

Reply:
It should check it in the "main" metod `Slice(start, length)`

Reply:
fair